### PR TITLE
Different Quickfix Behaviour Java vs. LSP #10112

### LINKED ISSUE #10112: Different Quickfix Behaviour Java vs. LSP
### Description
The difference is that in CHE-master we don't need to set the cursor into the position with a problem it is enough to set it into the line with a problem.
  
**Master**
qfmaster

**LSP**
qf_lsp

### PR DESCRIPTION
**Pull request by vrubezhny**
The fix makes LS Java Editor's quick fixes to follow CHE Java editor behavior

Fixes: #10112

Signed-off-by: Victor Rubezhny <vrubezhny@redhat.com>

### What does this PR do?

The fix makes LS Java Editor's quick fixes to follow CHE Java editor behavior

### What issues does this PR fix or reference?

Fixes: #10112

#### Release Notes

che10112

With this fix LSP Editor's quick fixes follow the CHE Java editor one's behavior

### CODE REVIEW

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: We should avoid leaving such comments.

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: Probably should be just `java.util.List`?

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: Instead of leaving a comment we can probably simply call the method `isInsideOurRange`. This will save us from unnecessary comment and at the same time will make the code declarative. What do you think?

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: I might be mistaken but why can't we have just 
```java
return start <= offset && offset <= end;
```
?

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java:201]: Now I see, you're using `isInside` method all over the pull request and give an explanation in comments to make it clearer for the reader. That is very nice of you however I would still recommend to avoid using comments in such cases and make the code itself more declarative instead.

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java:207]: However I admit that in some cases there is indeed no better way to explain the idea than to use a proper comment :-D

**Review by dkulieshov — APPROVED** [review]
Without taking into account minor remarks is an okay pull request.

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (dkulieshov) [inline plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: Probably it would be better to use `java.util.LinkedList` here and further as we don't know the exact size of a list, however I don't think this is so critical.

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (vrubezhny) [reply plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: isInside() method is an almost 1-to-1 copy from https://github.com/eclipse/che/blob/master/plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/editor/JavaQuickAssistProcessor.java#L313, so I wouldn't change its name here in order to 'not loose' the root from where this code came from.
As well as collectQuickFixableAnnotations() method, which is changed a lot due to the specifics of LSP DiagnosticAnnotation-s, but still a kind of copy of its namesake method from https://github.com/eclipse/che/blob/master/plugins/plugin-java/che-plugin-java-ext-lang-client/src/main/java/org/eclipse/che/ide/ext/java/client/editor/JavaQuickAssistProcessor.java#L192 as well as most of comments used in its body.

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (vrubezhny) [reply plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: I have already commented on the 'source' of the method, but I believe you're right and your replacement would work and looks better

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (vrubezhny) [reply plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java:201]: Again. Already commented on the source of these comments.

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (vrubezhny) [reply plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java:207]: In this case it is my own comment on that I wasn't able to replace JDT realization with something more or less similar in the LSP one. And as such this potentially can be a reason of why LSP realization works differently in some cases (and i believe there will be such cases)

File: plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java
Comment (vrubezhny) [reply plugins/plugin-languageserver/che-plugin-languageserver-ide/src/main/java/org/eclipse/che/plugin/languageserver/ide/editor/quickassist/LanguageServerQuickAssistProcessor.java]: I don't see it as something that can give a big improvement as in reality this list doesn't consist like hundreds of elements - in my case it has one element while there are four error annotations in the same line. We just need to store "more than one" diagnostics here as some error annotations might overlap... Not more, so it's not about the performance, imho.

**Comment by vrubezhny** [comment]
@dkuleshov I've updated PR according to your notes.
I've rebased it and, as such, pushed with force, so no fix-ups. Sorry about that.

**Comment by dkuleshov** [comment]
@vrubezhny excellent work thank you, we can merge this after we get it approved by the QE (@musienko-maxim).

**Comment by musienko-maxim** [comment]
We did not find regression in the E2E tests and during manual checking of LS features which is not covered by selenium tests

**Review by musienko-maxim — APPROVED** [review]

**Pull request merged**
